pontoon/static/js/tags/admin/index.js

```diff
@@ -0,0 +1,15 @@
+
+import React from 'react';
+import ReactDOM from 'react-dom';
+
+import TagResourcesButton from './button';
+
+
+document.addEventListener("DOMContentLoaded", () => {
+    document.querySelectorAll('.js-tag-resources').forEach(node => {
+        ReactDOM.render(
+            <TagResourcesButton
```

Comment:
This is kinda weird that this displays a button. It's also strange that the button has a data API (with `api` and all). I would expect there to be a father component here, that would handle the data, and then delegate showing a button and the interface or whatever. 

Not a blocker, but if you agree I'd love to see a follow-up to improve that.

Reply:
not sure what the best pattern is - these widgets are exceptional i think in that they are not part of an "app" - just inject to page.

im happy to code/review an update tho - but would need to be a bit clear about how it would work if im to code it

Reply:
If I understood the context properly (hard to tell without seeing something live), components need to be rendered in individual spread-out locations which have no common parent, because the surrounding layout is already rendered by server-side templates.

Unless the surrounding layout is also rendered via components, there will be no common parent, which means rendering via the top-level API (`ReactDOM.render()`) needs to be done N times, where N is the amount of `TagResourcesComponent` components to be rendered.

For passing the data to such components, reading it from the DOM as it's done here works, although IMO it's cleaner and more manageable to populate a data object in JS (i.e. in templates, initialize a JS object with relevant data), and have it read by glue code which is then responsible to render individual components. I have used this approach in the past and works good.

This pattern also allows to better handle situations where these components need to be re-rendered and/or added/removed due to interactions. Note as well that in those situations it's always important to unmount components to avoid keeping alive references to unused objects and event listeners.

Reply:
> For passing the data to such components, reading it from the DOM as it's done here works, although IMO it's cleaner and more manageable to populate a data object in JS

for the other UIs i worked on with react i did this - or mostly - there are some tradeoffs

in this case i dont think this is a good pattern - it would force the transfer of lots of data (list of all resources * n) that for the most part wouldnt get used

Reply:
In terms of data transfer, what I suggested is no different from the current code — again, if I understood the context correctly.

Reply:
in the current code the widgets only get filled - and data used - if the user clicks the button - they arent rendered (or at least the data is not called/rendered) at page load

Reply:
@julen maybe i misunderstand the point about passing it in DOM - in the current implementation the data is held as `state` in the dataManager hoc, and passed down as props. Are you suggesting storing it somewhere else after its retrieved from xhr ?

Reply:
What I was suggesting is that instead of storing the data in the DOM as data attributes, it can be rendered to a JS object, and use that to initialize a controller/parent module which serves as glue code and is responsible of doing the rendering of components (mounting/unmounting as needed).

Reply:
ah - i see - the api/project etc, rather than `data`

Reply:
IIRC, inline js data objects don't go well with CSP, or rather, are frowned upon from that school of thought.

Reply:
you can add `nonces` to the snippets - i think thats ok - see https://github.com/Bennyoak/django-csp-nonce

Reply:
although...

> This code has not been through a third party security audit.